src/network/mining_pool.rs

```diff
@@ -0,0 +1,324 @@
+            state: Arc::new(MiningPoolState::open_writer::<S, P>(path)?),
+            mining_pool_router,
+            memory_pool,
+            peers_router,
+            ledger_reader,
+            ledger_router,
+            prover_router,
+            current_template: RwLock::new(None),
+            miner_info: RwLock::new(HashMap::new()),
+        });
+
+        if E::NODE_TYPE == NodeType::MiningPool {
+            // Initialize the handler for the mining pool.
+            {
+                let mining_pool = mining_pool.clone();
+                let (router, handler) = oneshot::channel();
+                tasks.append(task::spawn(async move {
+                    // TODO: add loop which retargets share difficulty.
+                    // Notify the outer function that the task is ready.
+                    let _ = router.send(());
+                    // Asynchronously wait for a mining pool request.
+                    while let Some(request) = mining_pool_handler.recv().await {
+                        mining_pool.update(request).await;
+                    }
+                }));
+                // Wait until the mining pool handler is ready.
+                let _ = handler.await;
+            }
+
+            if let Some(recipient) = mining_pool_address {
+                // Set initial block template.
+                mining_pool.set_block_template(recipient).await?;
+            } else {
+                error!("Missing miner address. Please specify an Aleo address in order to run a mining pool");
+            }
+        }
+
+        Ok(mining_pool)
+    }
+
+    /// Returns an instance of the mining pool router.
+    pub fn router(&self) -> MiningPoolRouter<N> {
+        self.mining_pool_router.clone()
+    }
+
+    /// Returns all the shares in storage.
+    pub fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.state.to_shares()
+    }
+
+    ///
+    /// Performs the given `request` to the mining pool.
+    /// All requests must go through this `update`, so that a unified view is preserved.
+    ///
+    pub(super) async fn update(&self, request: MiningPoolRequest<N>) {
+        match request {
+            MiningPoolRequest::ProposedBlock(peer_ip, mut block, miner_address) => {
+                if let Some(current_template) = &*self.current_template.read().await {
+                    // Check that the block is relevant.
+                    if self.ledger_reader.latest_block_height().saturating_add(1) != block.height() {
+                        warn!("[ProposedBlock] Peer {} sent a stale candidate block.", peer_ip);
+                        return;
+                    }
+
+                    // Check that the block's coinbase transaction owner is the mining pool address.
+                    let records = match block.to_coinbase_transaction() {
+                        Ok(tx) => {
+                            let coinbase_records: Vec<Record<N>> = tx.to_records().collect();
```

Comment:
can this allocation via `collect()` be avoided? the records are immediately iterated over right afterwards

Reply:
Well, I do need to use the records later down the line. I can just retrieve the records again down the line but it'll be a bit more bulky on the code side.

Reply:
ah ok, I missed this :+1: